lib/extensions.js

```diff
@@ -106,9 +107,18 @@ module.exports = function(realmConstructor, environment) {
                     }
 
                     // recursively trigger `toJSON` for Realm instances with the same cache.
-                    result[key] = (value instanceof realmConstructor.Object || value instanceof realmConstructor.Collection)
-                        ? value.toJSON(key, cache)
-                        : value;
+                    if (value instanceof realmConstructor.Object || value instanceof realmConstructor.Collection) {
+                        result[key] = value.toJSON(key, cache);
+                    } else if (value instanceof realmConstructor.Dictionary) {
+                        // Dictionary special case to share the "cache" for dictionary-values,
+                        // in case of circular structures.
+                        result[key] =  Object.entries(value).reduce((total, [k, v]) => {
+                            total[k] = v instanceof realmConstructor.Object ? v.toJSON(k, cache) : v;
+                            return total;
+                        }, {});
```

Comment:
I think that will return an array of values instead of an object of key value pairs.  The unit tests should prove this

Reply:
> I think that will return an array of values instead of an object of key value pairs. 

That is what `Object.fromEntries()` is for. Except that it will throw because I am only returning the value, not the pair. Fixed: 

```js
                        result[key] =  Object.fromEntries(Object.entries(value).map(([k, v]) =>
                            [k, v instanceof realmConstructor.Object ? v.toJSON(k, cache) : v]
                        ));
```